### ISSUE TITLE
Catch Vue events in similar way to lifecycle hooks

### ISSUE THREAD

**Issue by DominikSerafin** [vuejs/vue#4628]
Hello 😄 

This is a proposal for a feature.

Basically it boils down to one thing. Catching Vue events triggered with `vm.$emit` in the same way you hook into lifecycle events.

http://jsfiddle.net/df4Lnuw6/92/

This fiddle should clearly show what I've in mind. I'm catching here `fooChanged` outside of methods and without using `vm.$on`. The listener is directly added to the Vue instance and is called everytime that event is triggered. I think it could also be described as a 'custom lifecycle event'.

From my perspective I think it fits nicely into the Vue current design and it'd be very convenient way to catch events. I've had multiple situations where something like this would make listening to bigger internal changes of a component a little cleaner.

What do you think?

**Comment by fnlctrl** [comment]
Hi, thanks for filling this issue.
This feature was in 1.x as the [`events`](http://v1.vuejs.org/api/#events) component option, and was then removed along with $dispatch and $broadcast in 2.0.

For now, custom events ($emit and $on) are meant for parent-child communication, and you can already declaratively use `<foo @bar="doSomething">` inside templates.

As for listening events from within a component, there's actually almost no practical benefit of adding a listener (e.g. with `vm.$on`) to catch an self-emitted 'event', since you can just call the method instead:

```js
new Vue({
    methods: {
       makeItBar: function(){
         this.foo = 'bar';
         this.$emit('fooChanged'); // $emit an event for parent, if applicable
         this.fooChanged() // simply call the method
       },
       fooChanged: function(){ ... },
    }
})
```

In other words, the custom events works best for parent-child communication. As for a component itself, using `this.$emit('foo', ...)` instead of just calling `this.foo(...)` is an anti-pattern. Therefore, an `events` option like in 1.x isn't really necessary.

**Comment by DominikSerafin** [comment]
@fnlctrl thanks for the comment, but the `events` from 1.x isn't exactly what I've had in mind. You have to nest it similarly to `vm.$on`. My proposal was about detecting the event on top level.

And yeah methods are the best solution in most cases but sometimes there's a "larger" event that would be kinda neat to treat almost like a lifecycle event. If such feature would be used right it could improve code readability.

**Comment by fnlctrl** [comment]
> but sometimes there's a "larger" event that would be kinda neat to treat almost like a lifecycle event. If such feature would be used right it could improve code readability.

I don't know exactly what your use case would be, but it sounds like you want to promote a method to an option (to improve code redability?), e.g.
```js
new Vue({
    // before
    methods: {
       handleMyImportantEvent: function(){ ... },
    }
    // after
    handleMyImportantEvent: function(){ ... },
})
```

you can access it with `this.$options.handleMyImportantEvent`, and then write a plugin/mixin to use it by adding lifecycle hooks
```
{
  created() {
    let myImportantEventHandler = this.$options.handleMyImportantEvent;
    If (myImportantEventHandler) {
      this.$on('foo', myImportantEventHandler)
    }
  },
  beforeDestroy() {
    let myImportantEventHandler = this.$options.handleMyImportantEvent;
    If (myImportantEventHandler) {
      this.$off('foo', myImportantEventHandler)
    }
  }
}
```

**Comment by yyx990803** [comment]
To be clear, magically inferring event listeners from top-level options are unlikely to be in core, but can be easily implemented via a userland plugin as @fnlctrl suggested.

**Comment by DominikSerafin** [comment]
@fnlctrl @yyx990803 thank you for the replies guys, it was something I was curious what other people would think about. And yeah - maybe it wouldn't have so much use cases after all.
